s3/pkg/service/multipart.go

```
@@ -0,0 +1,450 @@
+	}()
+
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+
+	bucket, err := s.MetaStorage.GetBucket(ctx, bucketName, true)
+	if err != nil {
+		log.Errorln("failed to get bucket from meta storage. err:", err)
+		return err
+	}
+
+	if !isAdmin {
+		switch bucket.Acl.CannedAcl {
+		case "public-read-write":
+			break
+		default:
+			if bucket.TenantId != tenantId {
+				log.Errorln("bucket owner is not equal to request owner.")
+				err = ErrBucketAccessForbidden
+				return err
+			}
+		}
+	}
+
+	attrs := in.Attrs
+	contentType, ok := attrs["Content-Type"]
+	if !ok {
+		contentType = "application/octet-stream"
+	}
+
+	backendName := bucket.DefaultLocation
+	backend, err := utils.GetBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+
+	sd, err := driver.CreateStorageDriver(backend.Type, backend)
+	if err != nil {
+		log.Errorln("failed to create storage. err:", err)
+		return err
+	}
+	res, err := sd.InitMultipartUpload(ctx, &pb.Object{BucketName: bucketName, ObjectKey: objectKey})
+	if err != nil {
+		log.Errorln("failed to init multipart upload. err:", err)
+		return err
+	}
+
+	multipartMetadata := MultipartMetadata{
+		InitiatorId:  tenantId,
+		TenantId:     bucket.TenantId,
+		UserId:       bucket.UserId,
+		ContentType:  contentType,
+		Acl:          datatype.Acl{CannedAcl: in.Acl.CannedAcl},
+		Attrs:        attrs,
+		StorageClass: StorageClass(in.StorageClass),
+	}
+
+	multipart := Multipart{
+		BucketName:  bucketName,
+		ObjectKey:   objectKey,
+		UploadId:    res.UploadId,
+		ObjectId:    res.ObjectId,
+		InitialTime: time.Now().UTC(),
+		Metadata:    multipartMetadata,
+	}
+
+	err = s.MetaStorage.Db.CreateMultipart(multipart)
+	if err != nil {
+		log.Errorln("failed to create multipart in meta. err:", err)
```

err is not handled here.

---

@sfzeng , Thanks, will do it